### ISSUE TITLE
IconLayer (png) render artifacts

### ISSUE THREAD

**Issue by jarekpelczynski** [visgl/deck.gl#2169]
I use IconLayer with png sprite.
Icons render with a small glitch, something like a gray border or shadow but originally they have only a white border.

![Preview](https://uc6bf0060c4f8e3a3fe9577b2ab8.previews.dropboxusercontent.com/p/thumb/AAIhfi2bMiZQ3JgDJMVtsKa_bdw3XwVs2YMWouquhHZbAKxfyEW11yZfXXEcCpIW-WHJdjhDQFeT2BozdvxuzrhEbQugT3huJu0x8MhXzQ7hWTYIvdUIpQKLeSFlVsusPf9bSJ3gPAfseE93Ld3OfKzLLapWnyvT23Dthd3KYtfIqp43t37OfDeXdi470QJiUlvzFZW4jiY05VqaaYX4tU_lAFLctMo6D_bXwcqIPqSi-w/p.png?size=1600x1200&size_mode=3)

**Comment by jianhuang01** [comment]
can you add your original png and the final rendering effects here?

**Comment by wooly** [comment]
hey @jianhuang01,

We've come across this issue as well, so hopefully this will help:

```
    const iconMapping = {
      default: {
        x: 0,
        y: 0,
        width: 80,
        height: 80,
      },
    }

    const iconLayer = new IconLayer({
      id: `icon-layer`,
      pickable: true,
      iconMapping,
      iconAtlas: 'locations-spritesheet-white.png',
      getIcon: () => 'default',
      getSize: 80,
      sizeScale: 5,
      data: layersData,
      opacity: 1,
    });
```
White sprite-sheet, only using the first sprite:
<img width="120" alt="locations-spritesheet-white" src="https://user-images.githubusercontent.com/351076/56263652-1f676780-6127-11e9-9b6d-195efb845ce0.png">

Rendered output in chrome:
<img width="354" alt="Screen Shot 2019-04-17 at 3 41 09 pm" src="https://user-images.githubusercontent.com/351076/56263688-4756cb00-6127-11e9-9356-b9c010d1d706.png">

When we add in the second sprite for `onHover`, the result looks like this:
<img width="368" alt="Screen Shot 2019-04-17 at 3 43 56 pm" src="https://user-images.githubusercontent.com/351076/56263775-a74d7180-6127-11e9-8514-9bd35f707e5b.png">

Any ideas?

**Comment by wooly** [comment]
You can also see it happening on the IconLayer deck.gl example here: https://deck.gl/#/examples/core-layers/icon-layer

It's noticeable when the markers overlap:
<img width="122" alt="Screen Shot 2019-04-17 at 4 19 55 pm" src="https://user-images.githubusercontent.com/351076/56265379-b5ea5780-612c-11e9-9561-2227747899c0.png">
<img width="151" alt="Screen Shot 2019-04-17 at 4 20 53 pm" src="https://user-images.githubusercontent.com/351076/56265403-cb5f8180-612c-11e9-95ad-ebdea2a71ee5.png">

Maybe something to do with alpha blending?

**Comment by tsherif** [comment]
@wooly Are those bitmaps using the alpha channel to fade off at the edges?
@jianhuang01 Does the icon layer use the alpha channel from the image it's sampling?

**Comment by wooly** [comment]
@tsherif they're using the alpha channel to anti-alias the circle edges, yep

**Comment by tsherif** [comment]
It looks like the icon layer does sample the texture alpha. The issue might be that deck.gl doesn't premultiply its output alphas (some discussion [here](https://webglfundamentals.org/webgl/lessons/webgl-and-alpha.html) of why this is an issue).

**Comment by tsherif** [comment]
@wooly Can you try adding the following to your icon layer constructor props:
```js
parameters: {
    depthMask: false
}
```

**Comment by wooly** [comment]
@tsherif didn't make any discernible difference, unfortunately.

**Comment by leeran88** [comment]
Happens also to me.
@tsherif , any workarounds?

**Comment by BowlingX** [comment]
Anything that can be done here? Happens for me as well with the latest version:

![image](https://user-images.githubusercontent.com/215671/87294098-88dd4e00-c503-11ea-85f3-d9627026b348.png)
![image](https://user-images.githubusercontent.com/215671/87294201-aad6d080-c503-11ea-89c0-474424dbf2aa.png)

```
    "@deck.gl/react": "^8.2.2",
    "@deck.gl/layers": "^8.2.2",
    "react-map-gl": "^5.2.7",
```

**Comment by BowlingX** [comment]
If I set `premultipliedAlpha: false`, the border get's a bit nicer

![image](https://user-images.githubusercontent.com/215671/87297348-8e896280-c508-11ea-829b-391b34e56ca0.png)

**Comment by ibgreen** [comment]
@BowlingX Good that you were able to pinpoint this. It is unexpected. [ImageBitmap](https://developer.mozilla.org/en-US/docs/Web/API/ImageBitmap) is the recommended format for loading images on modern browsers, but based on your findings it looks like there is potential issue with transparency when converted to texture. 

I googled a bit but could not find more information on this.  

For completeness, can you provide browser information?

If you have time, you could also try the following and see if it makes a difference:

```
new IconLayer({ ...yourProps, loadOptions: {
          imagebitmap: {
            premultiplyAlpha: 'premultiply'
          }
        }})
```

For information on that, see [createImageBitmap](https://developer.mozilla.org/en-US/docs/Web/API/WindowOrWorkerGlobalScope/createImageBitmap)

**Comment by BowlingX** [comment]
@ibgreen I tried this option already and it did not fix the issue, unfortunately. The only thing that helped was to change the type to `image`. Browser is Chrome  `83.0.4103.116`. But the same issue was also present on Firefox `77.0.1` and `78.0.2`. Operating system is Windows 10 with NVIDIA 2080 TI.

**Comment by Pessimistress** [comment]
@ibgreen I think we want `premultiplyAlpha: 'none'` to remove the border. I created a demo here: https://codepen.io/Pessimistress/pen/xxZQGMM

The option doesn't seem to have any effect either way in Chrome, but works in FireFox.

**Comment by ibgreen** [comment]
That sounds good. We could perhaps set it as a default option in the image loader?

https://github.com/visgl/loaders.gl/blob/master/modules/images/src/image-loader.js#L36

**Comment by Pessimistress** [comment]
@ibgreen ImageBitmap's options do not work in Chrome. Setting a default won't solve this issue.

**Comment by cneumann** [comment]
I've just run into this as well and I can confirm what @Pessimistress says about `loadOptions` for ImageBitmap not working on Chrome :(

When looking at what Chrome renders with [renderdoc](https://renderdoc.org) I can see that the texture holding the icons is indeed premultiplied (e.g. fully transparent pixels have value RGBA: 0x00000000).
However, the blend function set when drawing is: `GL_SRC_ALPHA`, `GL_ONE_MINUS_SRC_ALPHA`; I believe it should be `GL_ONE`, `GL_ONE_MINUS_SRC_ALPHA` for premultiplied texture data?

I think [icon-layer.js#L208](https://github.com/visgl/deck.gl/blob/d53aaabb81bd8e9894f92e26648b05cfb746d907/modules/layers/src/icon-layer/icon-layer.js#L208) should call draw with custom parameters, e.g. something like:

```javascript
    this.state.model
        .setUniforms(
          Object.assign({}, uniforms, {
            iconsTexture,
            iconsTextureDim: [iconsTexture.width, iconsTexture.height],
            sizeScale: sizeScale * (sizeUnits === 'pixels' ? viewport.metersPerPixel : 1),
            sizeMinPixels,
            sizeMaxPixels,
            billboard,
            alphaCutoff
          })
        )
        .draw({
            parameters: {
                blendFunc: [GL.ONE, GL.ONE_MINUS_SRC_ALPHA]
          }
        });
```

**Comment by Pessimistress** [comment]
Thanks @cneumann for the insight!

Can confirm that this solves the problem:

```js
  new IconLayer({
    ...
    parameters: {
        blendFunc: [1, 771] // GL.ONE, GL.ONE_MINUS_SRC_ALPHA
    },
    loadOptions: {
        imagebitmap: {
          premultiplyAlpha: 'none'
        }
    }
})
```

We can change `IconLayer`'s default blending parameters, though that could be a breaking change for users who are customizing global blending. @ibgreen

**Comment by cneumann** [comment]
Just to be sure: if you change the blend function in this way you *do* want premultiplied alpha, so the option should be `imagebitmap.premultiplyAlpha = 'premultiply'` - of course that only matters when/if the options actually get passed to the browser ;)

**Comment by Pessimistress** [comment]
Hmm, you are correct; although changing the option does not seem to make any difference. It looks like Chrome's bitmap color is always premultiplied regardless.

I can confirm the the option is actually passed to `createImageBitmap`.

**Comment by ibgreen** [comment]
Ok, this seems to be a rabbithole. I suppose ImageBitmaps are supposed to be in a format that is particularly efficient for rendering, maybe the implementation is allowed this freedom.

Maybe try loading as Image instead of ImageBitmap then. Less performant but correct results... Something along these lines?

```js
new IconLayer({ ...yourProps, loadOptions: {
          image: {
            type: 'image'
          }
        }})
```

**Comment by Pessimistress** [comment]
Edit: no, the option is not passed to `createImageBitmap`. This is a bug in loaders.gl. I'll put up a fix.

**Comment by Pessimistress** [comment]
Update: an issue is fixed in `@loaders.gl/images@2.2.9`, and the artifacts can be removed with

```js
new IconLayer({
    ...
    loadOptions: {
        imagebitmap: {
          premultiplyAlpha: 'none'
        }
    }
})
```

I plan to make `premultiplyAlpha: 'none'` the default in the upcoming 8.3.0 release so that this workaround will no longer be needed.

**Comment by Pessimistress** [comment]
Fixed in 8.3.0-beta.2
